File: chapter4 - Trees and Graphs/4.4_checkBalanced/Java/efficient/BST.java
package efficient;

import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;

public class BST {
  public TreeNode root;

  public BST(int val) {
    this.root = new TreeNode(val);
  }

  public void insert(int val) {
    this.root.insert(val);
  }

  public void printLevel() {
    ArrayList<Integer> level = new ArrayList<>();
    Queue<TreeNode> q = new LinkedList<>();
    Queue<TreeNode> nextq = new LinkedList<>();

    q.add(this.root);

    while (!q.isEmpty()) {
      TreeNode currNode = q.poll();
      level.add(currNode.val);

      if (currNode.left != null) {
        nextq.add(currNode.left);
      }

      if (currNode.right != null) {
        nextq.add(currNode.right);
      }

      if (q.isEmpty()) {
        System.out.println(level);
        q = nextq;
        nextq = new LinkedList<>();
        level = new ArrayList<>();
      }
    }
  }
}

File: chapter4 - Trees and Graphs/4.4_checkBalanced/JavaScript/BST.js
const TreeNode = require("./TreeNode");

class BST {
  constructor(val) {
    this.root = new TreeNode(val);
  }

  insert(val) {
    this.root.insert(val);
  }

  printLevel() {
    let q = [];
    let nextq = [];
    let level = []

    q.push(this.root);
    
    while (q.length) {
      const currNode = q.shift();
      level.push(currNode.val);

      if (currNode.left) {
        nextq.push(currNode.left);
      }

      if (currNode.right) {
        nextq.push(currNode.right);
      }

      if (!q.length) {
        console.log(level);
        q = [...nextq];
        nextq = [];
        level = [];
      }
    }
  }
}

module.exports = BST;